File: rental_service.py
from model import Rent, Status
from datetime import date


class RentalService:
    def __init__(self, item_service):
        self.rentals = []
        self.item_service = item_service

    def rent_item(self, scanner):
        item_id = input("Enter item ID to rent: ")
        item = self.item_service.find_by_id(item_id)

        if item is None:
            print("Item not found.")
            return

        if item.status == Status.DISABLED:
            print("Sorry, item is disabled")
            return

        start_date = self.prompt_date("Enter start date (YYYY-MM-DD): ")
        end_date = self.prompt_date("Enter end date (YYYY-MM-DD): ")

        if end_date <= start_date:
            print("Rental must be at least 1 day long.")
            return

        if date.today() > start_date:
            print("Rental must be in the future")
            return

        for rent in self.rentals:
            if rent.item_for_rent.id == item_id and self.is_between_rent_days(start_date, end_date, rent):
                print(f"Item is already rented from {rent.start_date} to {rent.end_date}")
                return

        duration = (end_date - start_date).days
        rent = Rent(start_date, duration, item)
        item.status = Status.RESERVED
        self.rentals.append(rent)
        print("Rental successful:", rent)

    def is_between_rent_days(self, start_date, end_date, rent: Rent):
        rent_start = rent.start_date
        rent_end = rent.end_date
        return rent_start <= end_date and rent_end >= start_date

    def list_active_rentals(self):
        today = date.today()
        print("\nActive Rentals:")
        active = [r for r in self.rentals if r.end_date > today]
        for r in active:
            print(f"{r.item_for_rent.name} rented from {r.start_date} to {r.end_date} (${r.price:.2f})")
        return active

    def prompt_date(self, prompt: str):
        while True:
            try:
                return date.fromisoformat(input(prompt))
            except ValueError:
                print("Invalid date format. Use YYYY-MM-DD.")

File: RentalService.js
const Rent = require("./Rent");
const Status = require("./Status");

class RentalService {
  constructor(itemService) {
    this.itemService = itemService;
    this.rentals = [];
  }

  rentItem(rl, callback) {
    rl.question("Enter item ID to rent: ", itemId => {
      const item = this.itemService.findById(itemId);
      if (!item) {
        console.log("Item not found.");
        return callback();
      }
      if (item.status === Status.DISABLED) {
        console.log("Sorry, item is disabled");
        return callback();
      }

      this.promptDate(rl, "Enter start date (YYYY-MM-DD): ", startDate => {
        this.promptDate(rl, "Enter end date (YYYY-MM-DD): ", endDate => {
          const start = new Date(startDate);
          const end = new Date(endDate);
          if (end <= start) {
            console.log("Rental must be at least 1 day long.");
            return callback();
          }
          const now = new Date();
          if (start <= now) {
            console.log("Rental must be in the future");
            return callback();
          }

          for (const r of this.rentals) {
            if (
              r.itemForRent.id === itemId &&
              this.isBetweenRentDays(start, end, r)
            ) {
              console.log(`Item is already rented from ${r.startDate.toISOString().split('T')[0]} to ${r.endDate.toISOString().split('T')[0]}`);
              return callback();
            }
          }

          const duration = Math.round((end - start) / (1000 * 60 * 60 * 24));
          const rent = new Rent(startDate, duration, item);
          item.status = Status.RESERVED;
          this.rentals.push(rent);
          console.log("Rental successful: " + rent.toString());
          callback();
        });
      });
    });
  }

  isBetweenRentDays(startDate, endDate, rent) {
    return rent.startDate <= endDate && rent.endDate >= startDate;
  }

  listActiveRentals() {
    const today = new Date();
    console.log("\nActive Rentals:");
    const result = this.rentals.filter(r => r.endDate > today);
    for (const r of result) {
      console.log(`${r.itemForRent.name} rented from ${r.startDate.toISOString().split('T')[0]} to ${r.endDate.toISOString().split('T')[0]} ($${r.price.toFixed(2)})`);
    }
    return result;
  }

  promptDate(rl, prompt, callback) {
    rl.question(prompt, date => {
      if (!/^\\d{4}-\\d{2}-\\d{2}$/.test(date)) {
        console.log("Invalid date format. Use YYYY-MM-DD.");
        return this.promptDate(rl, prompt, callback);
      }
      callback(date);
    });
  }
}

module.exports = RentalService;
